mldsa/native/aarch64/src/intt.S

```diff
@@ -0,0 +1,409 @@
+
+        // layer4
+        gs_butterfly data0, data1, root3, 2, 3
+        gs_butterfly data2, data3, root4, 0, 1
+        gs_butterfly data4, data5, root4, 2, 3
+        gs_butterfly data6, data7, root5, 0, 1
+        gs_butterfly data8, data9, root5, 2, 3
+        gs_butterfly data10, data11, root6, 0, 1
+        gs_butterfly data12, data13, root6, 2, 3
+        gs_butterfly data14, data15, root7, 0, 1
+
+        // layer3
+        gs_butterfly data0, data2, root1, 2, 3
+        gs_butterfly data1, data3, root1, 2, 3
+        gs_butterfly data4, data6, root2, 0, 1
+        gs_butterfly data5, data7, root2, 0, 1
+        gs_butterfly data8, data10, root2, 2, 3
+        gs_butterfly data9, data11, root2, 2, 3
+        gs_butterfly data12, data14, root3, 0, 1
+        gs_butterfly data13, data15, root3, 0, 1
+
+        // layer2
+        gs_butterfly data0, data4, root0, 2, 3
+        gs_butterfly data1, data5, root0, 2, 3
+        gs_butterfly data2, data6, root0, 2, 3
+        gs_butterfly data3, data7, root0, 2, 3
+        gs_butterfly data8, data12, root1, 0, 1
+        gs_butterfly data9, data13, root1, 0, 1
+        gs_butterfly data10, data14, root1, 0, 1
+        gs_butterfly data11, data15, root1, 0, 1
+
+        // layer 1
+        gs_butterfly data0, data8, root0, 0, 1
+        gs_butterfly data1, data9, root0, 0, 1
+        gs_butterfly data2, data10, root0, 0, 1
+        gs_butterfly data3, data11, root0, 0, 1
+        gs_butterfly data4, data12, root0, 0, 1
+        gs_butterfly data5, data13, root0, 0, 1
+        gs_butterfly data6, data14, root0, 0, 1
+        gs_butterfly data7, data15, root0, 0, 1
+
+        canonical_reduce data8,  modulus_half, neg_modulus_half, t2, t3
+        canonical_reduce data9,  modulus_half, neg_modulus_half, t2, t3
+        canonical_reduce data10, modulus_half, neg_modulus_half, t2, t3
+        canonical_reduce data11, modulus_half, neg_modulus_half, t2, t3
+        canonical_reduce data12, modulus_half, neg_modulus_half, t2, t3
+        canonical_reduce data13, modulus_half, neg_modulus_half, t2, t3
+        canonical_reduce data14, modulus_half, neg_modulus_half, t2, t3
+        canonical_reduce data15, modulus_half, neg_modulus_half, t2, t3
+
+        str q_data8,  [in,  #(8*(512/8))]
+        str q_data9,  [in,  #(9*(512/8))]
+        str q_data10, [in,  #(10*(512/8))]
+        str q_data11, [in,  #(11*(512/8))]
+        str q_data12, [in,  #(12*(512/8))]
+        str q_data13, [in,  #(13*(512/8))]
+        str q_data14, [in,  #(14*(512/8))]
+        str q_data15, [in,  #(15*(512/8))]
+
+        // Scale half the coeffs by 1/n; for the other half, the scaling has
+        // been merged into the multiplication with the twiddle factor on the
+        // last layer.
+        mul_ninv data0, data1, data2, data3, data4, data5, data6, data7, data0, data1, data2, data3, data4, data5, data6, data7
+
+        canonical_reduce data0, modulus_half, neg_modulus_half, t2, t3
```

Comment:
I don't think we need a canonical reduction here, do we? The caller conducts further processing on the output of the invNTT which necessitates a canonical reduction at a later point.

Reply:
Agreed. That was an artifact from the paper so that we could compare to a previous implementation.